File: fastlowpointer/287.FindtheDuplicateNumber.go
package main

import "fmt"

func findDuplicate(nums []int) int {
	slow := 0
	fast := 0
	slow = nums[slow]
	fast = nums[fast]
	fast = nums[fast]
	for fast != slow {
		slow = nums[slow]
		fast = nums[fast]
		fast = nums[fast]
	}
	anther := 0
	for anther != slow {
		anther = nums[anther]
		slow = nums[slow]
	}
	return slow
}

func main() {
	nums := []int{1, 3, 4, 2, 2} // 2
	fmt.Println(findDuplicate(nums))
}

File: fastlowpointer/287.FindtheDuplicateNumber.py
from typing import List


class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        fast = slow = 0
        slow = nums[slow]
        fast = nums[fast]
        fast = nums[fast]
        while slow != fast:
            slow = nums[slow]
            fast = nums[fast]
            fast = nums[fast]
        slow = 0
        while slow != fast:
            slow = nums[slow]
            fast = nums[fast]
        return slow


if __name__ == "__main__":

    s = Solution()
    l = [1, 3, 4, 2, 2]
    print(s.findDuplicate(l))
    l = [3, 1, 3, 4, 2]
    print(s.findDuplicate(l))
